File: svc-data/src/test/java/com/beancounter/marketdata/integ/MarketDataControllerTests.java
package com.beancounter.marketdata.integ;

import static com.beancounter.common.utils.AssetUtils.getAssetInput;
import static org.assertj.core.api.Assertions.assertThat;
import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.get;
import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.post;
import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.content;
import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.status;

import com.beancounter.auth.server.RoleHelper;
import com.beancounter.common.contracts.MarketResponse;
import com.beancounter.common.contracts.PriceRequest;
import com.beancounter.common.contracts.PriceResponse;
import com.beancounter.common.input.AssetInput;
import com.beancounter.common.model.Asset;
import com.beancounter.common.model.MarketData;
import com.beancounter.common.utils.AssetUtils;
import com.beancounter.common.utils.BcJson;
import com.beancounter.marketdata.MarketDataBoot;
import com.beancounter.marketdata.assets.AssetService;
import com.beancounter.marketdata.markets.MarketService;
import com.beancounter.marketdata.providers.mock.MockProviderService;
import com.fasterxml.jackson.core.JsonProcessingException;
import java.math.BigDecimal;
import java.util.ArrayList;
import java.util.Collection;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Tag;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.Mockito;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.context.SpringBootTest;
import org.springframework.boot.test.mock.mockito.MockBean;
import org.springframework.http.MediaType;
import org.springframework.security.test.context.support.WithMockUser;
import org.springframework.test.context.ActiveProfiles;
import org.springframework.test.context.junit.jupiter.SpringExtension;
import org.springframework.test.context.web.WebAppConfiguration;
import org.springframework.test.web.servlet.MockMvc;
import org.springframework.test.web.servlet.setup.MockMvcBuilders;
import org.springframework.web.context.WebApplicationContext;

@ExtendWith(SpringExtension.class)
@SpringBootTest(classes = MarketDataBoot.class)
@WebAppConfiguration
@ActiveProfiles("test")
class MarketDataControllerTests {

  private final Asset dummy;
  // Represents dummy after it has been Jackson'ized
  private final Asset dummyJsonAsset;
  private final WebApplicationContext wac;
  private final MockProviderService mockProviderService;
  @MockBean
  private AssetService assetService;
  private MockMvc mockMvc;
  private final BcJson bcJson;

  @Autowired
  private MarketDataControllerTests(WebApplicationContext webApplicationContext,
                                    MarketService marketService,
                                    MockProviderService mockProviderService, BcJson bcJson)
      throws JsonProcessingException {
    this.wac = webApplicationContext;
    this.mockProviderService = mockProviderService;
    dummy = AssetUtils.getAsset(
        marketService.getMarket("mock"), "dummy"
    );
    this.bcJson = bcJson;
    dummyJsonAsset = this.bcJson.getObjectMapper()
        .readValue(this.bcJson.getObjectMapper().writeValueAsString(dummy), Asset.class);
  }

  @BeforeEach
  void setUp() {
    this.mockMvc = MockMvcBuilders.webAppContextSetup(this.wac).build();
    assertThat(dummy.getId()).isNotNull();
    Mockito.when(
        assetService.find(dummy.getId())).thenReturn(dummy);
    assertThat(dummy.getId()).isNotNull();
    Mockito.when(
        assetService.findLocally(dummy.getMarket().getCode(), dummy.getCode()))
        .thenReturn(dummy);
  }

  @Test
  void is_ContextLoaded() {
    assertThat(wac).isNotNull();
  }

  @Test
  @WithMockUser(username = "test-user", roles = {RoleHelper.OAUTH_USER})
  void is_MarketsReturned() throws Exception {
    String json = mockMvc.perform(get("/markets")
        .contentType(MediaType.APPLICATION_JSON_VALUE))
        .andExpect(
            status().isOk()
        ).andExpect(
            content().contentType(MediaType.APPLICATION_JSON_VALUE)
        ).andReturn()
        .getResponse().getContentAsString();
    MarketResponse marketResponse = bcJson.getObjectMapper().readValue(json, MarketResponse.class);
    assertThat(marketResponse.getData()).isNotNull().isNotEmpty();
  }

  @Test
  @Tag("slow")
  @WithMockUser(username = "test-user", roles = {RoleHelper.OAUTH_USER})
  void is_PriceFormMarketAssetFound() throws Exception {

    String json = mockMvc.perform(
        get("/prices/{marketId}/{assetId}",
            dummy.getMarket().getCode(),
            dummy.getCode()
        )
            .contentType(MediaType.APPLICATION_JSON_VALUE))
        .andExpect(
            status().isOk()
        ).andExpect(
            content().contentType(MediaType.APPLICATION_JSON_VALUE)
        ).andReturn().getResponse().getContentAsString();

    PriceResponse priceResponse = bcJson.getObjectMapper().readValue(json, PriceResponse.class);

    assertThat(priceResponse.getData()).isNotNull().hasSize(1);
    MarketData marketData = priceResponse.getData().iterator().next();
    assertThat(marketData)
        .hasFieldOrPropertyWithValue("asset", dummyJsonAsset)
        .hasFieldOrPropertyWithValue("open", BigDecimal.valueOf(999.99))
        .hasFieldOrPropertyWithValue("priceDate", mockProviderService.getPriceDate());
  }

  @Test
  @Tag("slow")
  @WithMockUser(username = "test-user", roles = {RoleHelper.OAUTH_USER})
  void is_MdCollectionReturnedForAssets() throws Exception {

    Mockito.when(assetService.findLocally("MOCK", "ASSETCODE"))
        .thenReturn(AssetUtils.getAsset("ASSETCODE", "MOCK"));

    Collection<AssetInput> assetInputs = new ArrayList<>();
    assetInputs.add(
        getAssetInput("MOCK", "ASSETCODE"));

    String json = mockMvc.perform(
        post("/prices")
            .contentType(MediaType.APPLICATION_JSON_VALUE)
            .content(bcJson.getObjectMapper().writeValueAsString(
                new PriceRequest(assetInputs))
            ))
        .andExpect(
            status().isOk()
        ).andExpect(
            content().contentType(MediaType.APPLICATION_JSON_VALUE)
        ).andReturn()
        .getResponse()
        .getContentAsString();

    PriceResponse mdResponse = bcJson.getObjectMapper().readValue(json, PriceResponse.class);
    assertThat(mdResponse.getData())
        .isNotNull()
        .hasSize(assetInputs.size());
  }

  @Test
  @Tag("slow")
  @WithMockUser(username = "test-user", roles = {RoleHelper.OAUTH_USER})
  void is_ValuationRequestHydratingAssets() throws Exception {
    String json = mockMvc.perform(get("/prices/{marketId}/{assetId}",
        dummy.getMarket().getCode(),
        dummy.getCode())
        .contentType(MediaType.APPLICATION_JSON_VALUE)
    ).andExpect(
        status().isOk()
    ).andExpect(
        content().contentType(MediaType.APPLICATION_JSON_VALUE)
    ).andReturn()
        .getResponse()
        .getContentAsString();

    PriceResponse priceResponse = bcJson.getObjectMapper().readValue(json, PriceResponse.class);

    assertThat(priceResponse.getData()).isNotNull().hasSize(1);

    MarketData marketData = priceResponse.getData().iterator().next();
    assertThat(marketData)
        .hasFieldOrPropertyWithValue("asset", dummyJsonAsset)
        .hasFieldOrPropertyWithValue("open", BigDecimal.valueOf(999.99))
        .hasFieldOrPropertyWithValue("priceDate", mockProviderService.getPriceDate());


  }
}

File: svc-data/src/test/java/com/beancounter/marketdata/integ/MarketDataControllerTests.kt
package com.beancounter.marketdata.integ

import com.beancounter.auth.server.RoleHelper
import com.beancounter.common.contracts.MarketResponse
import com.beancounter.common.contracts.PriceRequest
import com.beancounter.common.contracts.PriceResponse
import com.beancounter.common.input.AssetInput
import com.beancounter.common.model.Asset
import com.beancounter.common.utils.AssetUtils.Companion.getAsset
import com.beancounter.common.utils.AssetUtils.Companion.getAssetInput
import com.beancounter.common.utils.BcJson
import com.beancounter.marketdata.assets.AssetService
import com.beancounter.marketdata.markets.MarketService
import com.beancounter.marketdata.providers.mock.MockProviderService
import org.assertj.core.api.Assertions
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Tag
import org.junit.jupiter.api.Test
import org.mockito.Mockito
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.boot.autoconfigure.domain.EntityScan
import org.springframework.boot.test.context.SpringBootTest
import org.springframework.boot.test.mock.mockito.MockBean
import org.springframework.http.MediaType
import org.springframework.security.test.context.support.WithMockUser
import org.springframework.test.context.ActiveProfiles
import org.springframework.test.web.servlet.MockMvc
import org.springframework.test.web.servlet.request.MockMvcRequestBuilders
import org.springframework.test.web.servlet.result.MockMvcResultMatchers
import org.springframework.test.web.servlet.setup.MockMvcBuilders
import org.springframework.web.context.WebApplicationContext
import java.math.BigDecimal
import java.util.ArrayList

@SpringBootTest
@ActiveProfiles("test")
@Tag("slow")
@EntityScan("com.beancounter.common.model")
internal class MarketDataControllerTests @Autowired private constructor(
    private val wac: WebApplicationContext,
    marketService: MarketService,
    private val mockProviderService: MockProviderService,
    private val bcJson: BcJson,
) {
    private val dummy: Asset = getAsset(
        marketService.getMarket("mock"), "dummy"
    )

    // Represents dummy after it has been Jackson'ized
    private val dummyJsonAsset: Asset = this.bcJson.objectMapper
        .readValue(this.bcJson.objectMapper.writeValueAsString(dummy), Asset::class.java)

    @MockBean
    private lateinit var assetService: AssetService
    private lateinit var mockMvc: MockMvc

    @BeforeEach
    fun setUp() {
        mockMvc = MockMvcBuilders.webAppContextSetup(wac).build()
        Assertions.assertThat(dummy.id).isNotNull
        Mockito.`when`(
            assetService.find(dummy.id)
        ).thenReturn(dummy)
        Mockito.`when`(
            assetService.findLocally(dummy.market.code, dummy.code)
        )
            .thenReturn(dummy)
    }

    @Test
    fun is_ContextLoaded() {
        Assertions.assertThat(wac).isNotNull
    }

    @Test
    @WithMockUser(username = "test-user", roles = [RoleHelper.OAUTH_USER])
    @Throws(
        Exception::class
    )
    fun is_MarketsReturned() {
        val json = mockMvc.perform(
            MockMvcRequestBuilders.get("/markets")
                .contentType(MediaType.APPLICATION_JSON_VALUE)
        )
            .andExpect(
                MockMvcResultMatchers.status().isOk
            ).andExpect(
                MockMvcResultMatchers.content().contentType(MediaType.APPLICATION_JSON_VALUE)
            ).andReturn()
            .response.contentAsString
        val (data) = bcJson.objectMapper.readValue(json, MarketResponse::class.java)
        Assertions.assertThat(data).isNotNull.isNotEmpty
    }

    @Test
    @Tag("slow")
    @WithMockUser(username = "test-user", roles = [RoleHelper.OAUTH_USER])
    @Throws(
        Exception::class
    )
    fun is_PriceFormMarketAssetFound() {
        val json = mockMvc.perform(
            MockMvcRequestBuilders.get(
                "/prices/{marketId}/{assetId}",
                dummy.market.code,
                dummy.code
            )
                .contentType(MediaType.APPLICATION_JSON_VALUE)
        )
            .andExpect(
                MockMvcResultMatchers.status().isOk
            ).andExpect(
                MockMvcResultMatchers.content().contentType(MediaType.APPLICATION_JSON_VALUE)
            ).andReturn().response.contentAsString
        val (data) = bcJson.objectMapper.readValue(json, PriceResponse::class.java)
        Assertions.assertThat(data).isNotNull.hasSize(1)
        val marketData = data.iterator().next()
        Assertions.assertThat(marketData)
            .hasFieldOrPropertyWithValue("asset", dummyJsonAsset)
            .hasFieldOrPropertyWithValue("open", BigDecimal.valueOf(999.99))
            .hasFieldOrPropertyWithValue("priceDate", mockProviderService.priceDate)
    }

    @Test
    @Tag("slow")
    @WithMockUser(username = "test-user", roles = [RoleHelper.OAUTH_USER])
    @Throws(
        Exception::class
    )
    fun is_MdCollectionReturnedForAssets() {
        Mockito.`when`(assetService.findLocally("MOCK", "ASSETCODE"))
            .thenReturn(getAsset("ASSETCODE", "MOCK"))
        val assetInputs: MutableCollection<AssetInput> = ArrayList()
        assetInputs.add(
            getAssetInput("MOCK", "ASSETCODE")
        )
        val json = mockMvc.perform(
            MockMvcRequestBuilders.post("/prices")
                .contentType(MediaType.APPLICATION_JSON_VALUE)
                .content(
                    bcJson.objectMapper.writeValueAsString(
                        PriceRequest(assetInputs)
                    )
                )
        )
            .andExpect(
                MockMvcResultMatchers.status().isOk
            ).andExpect(
                MockMvcResultMatchers.content().contentType(MediaType.APPLICATION_JSON_VALUE)
            ).andReturn()
            .response
            .contentAsString
        val (data) = bcJson.objectMapper.readValue(json, PriceResponse::class.java)
        Assertions.assertThat(data)
            .isNotNull
            .hasSize(assetInputs.size)
    }

    @Test
    @Tag("slow")
    @WithMockUser(username = "test-user", roles = [RoleHelper.OAUTH_USER])
    @Throws(
        Exception::class
    )
    fun is_ValuationRequestHydratingAssets() {
        val json = mockMvc.perform(
            MockMvcRequestBuilders.get(
                "/prices/{marketId}/{assetId}",
                dummy.market.code,
                dummy.code
            )
                .contentType(MediaType.APPLICATION_JSON_VALUE)
        ).andExpect(
            MockMvcResultMatchers.status().isOk
        ).andExpect(
            MockMvcResultMatchers.content().contentType(MediaType.APPLICATION_JSON_VALUE)
        ).andReturn()
            .response
            .contentAsString
        val (data) = bcJson.objectMapper.readValue(json, PriceResponse::class.java)
        Assertions.assertThat(data).isNotNull.hasSize(1)
        val marketData = data.iterator().next()
        Assertions.assertThat(marketData)
            .hasFieldOrPropertyWithValue("asset", dummyJsonAsset)
            .hasFieldOrPropertyWithValue("open", BigDecimal.valueOf(999.99))
            .hasFieldOrPropertyWithValue("priceDate", mockProviderService.priceDate)
    }
}
